table/block_based/block_based_table_builder.cc

```diff
@@ -263,18 +229,42 @@ struct BlockBasedTableBuilder::Rep {
 
   std::string last_ikey;  // Internal key or empty (unset)
   const Slice* first_key_in_next_block = nullptr;
-  CompressionType compression_type;
+
   uint64_t sample_for_compression;
   std::atomic<uint64_t> compressible_input_data_bytes;
   std::atomic<uint64_t> uncompressible_input_data_bytes;
   std::atomic<uint64_t> sampled_input_data_bytes;
   std::atomic<uint64_t> sampled_output_slow_data_bytes;
   std::atomic<uint64_t> sampled_output_fast_data_bytes;
-  CompressionOptions compression_opts;
-  std::unique_ptr<CompressionDict> compression_dict;
-  std::vector<std::unique_ptr<CompressionContext>> compression_ctxs;
-  std::vector<std::unique_ptr<UncompressionContext>> verify_ctxs;
-  std::unique_ptr<UncompressionDict> verify_dict;
+  uint32_t compression_parallel_threads;
+  int max_compressed_bytes_per_kb;
+  size_t max_dict_sample_bytes = 0;
+  // A compressor for blocks in general, without dictionary compression
+  std::unique_ptr<Compressor> basic_compressor;
+  // A compressor using dictionary compression (when applicable)
+  std::unique_ptr<Compressor> compressor_with_dict;
+  // Once configured/determined, points to one of the above Compressors to
+  // use on data blocks.
+  Compressor* data_block_compressor = nullptr;
+  // A decompressor corresponding to basic_compressor (when non-nullptr).
+  // Used for verification and cache warming.
+  std::shared_ptr<Decompressor> basic_decompressor;
+  // When needed, a decompressor for verifying compression using a
+  // dictionary sampled/trained from this file.
+  std::unique_ptr<Decompressor> verify_decompressor_with_dict;
+  // When non-nullptr, compression should be verified with this corresponding
+  // decompressor, except for data blocks. (Points to same as basic_decompressor
+  // when verify_compression is set.)
+  UnownedPtr<Decompressor> verify_decompressor;
+  // Once configured/determined, points to one of the above Decompressors to use
+  // in verifying data blocks.
+  UnownedPtr<Decompressor> data_block_verify_decompressor;
```

Comment:
nit: it flows better from basic_decompressor -> verify_decompressor -> verify_decompressor_with_dict -> data_block_verify_decompressor since verify_decompressor will never be verify_decompressor_with_dict?